src/coroutine.h

```diff
@@ -0,0 +1,265 @@
+    void return_value(T&& value)
+    {
+      value_ = std::move(value);
+      auto awaitingCoroutine = awaitingCoroutine_;
+      if (awaitingCoroutine) {
+        awaitingCoroutine_ = nullptr;
+        std::coroutine_handle<>::from_address(awaitingCoroutine).resume();
+      }
+    }
+    T value_{};
+    void* awaitingCoroutine_ = nullptr;
+    bool earlyResume_ = false;
+    bool suspended_ = false;
+  };
+  typedef
+      typename std::conditional_t<std::is_void_v<T>, PromiseVoid, PromiseValue>
+          Promise;
+
+ public:
+  // The SafeT alias is used to handle the case where T is void. In this case,
+  // we use std::monostate as the type of the value_ member of the Promise
+  // class. This allows us to use a single Coroutine class for both void and
+  // value-returning coroutines.
+  typedef typename std::conditional_t<std::is_void_v<T>, std::monostate, T>
+      SafeT;
+
+  Coroutine() = default;
+  Coroutine(Coroutine&& other) = default;
+  Coroutine& operator=(Coroutine&& other) = default;
+  // The copy constructor and copy assignment operator are deleted copying
+  // coroutines doesn't make sense.
+  Coroutine(Coroutine const&) = delete;
+  Coroutine& operator=(Coroutine const&) = delete;
+
+  /**
+   * @brief A helper class used to implement the coroutine awaitable interface.
+   *
+   * @details The Awaiter class provides the necessary methods for the co_await
+   * operator to work with coroutines. Its interface is based on the C++20
+   * coroutine specification, and the names of the methods cannot be changed.
+   *
+   * This class enables proper synchronization between coroutines and allows for
+   * optimization through early resume functionality to avoid unnecessary
+   * suspensions.
+   *
+   * Its main purpose is to be a generic awaiter mechanism, in case creating a
+   * specific awaiter for an asynchronous operation would be too cumbersome, but
+   * in general, it is recommended to create a specific awaiter for each
+   * asynchronous operation instead.
+   */
+  class Awaiter {
+   public:
+    Awaiter(Awaiter&& other) = default;
+    Awaiter& operator=(Awaiter&& other) = default;
+    Awaiter(Awaiter const&) = default;
+    Awaiter& operator=(Awaiter const&) = default;
+    constexpr bool await_ready() const noexcept { return false; }
+    template <typename U>
+    constexpr bool await_suspend(std::coroutine_handle<U> h)
+    {
+      auto& promise = h.promise();
+      bool ret = promise.earlyResume_;
+      promise.earlyResume_ = false;
+      if (!ret) {
+        promise.suspended_ = true;
+      }
+      return !ret;
+    }
+    constexpr void await_resume() const noexcept {}
+  };
+
+  /**
+   * @brief Resumes the coroutine.
+   *
+   * @details This method resumes the coroutine if it is suspended using the
+   * Awaiter mechanism above, or for its initial execution. If the coroutine is
+   * not suspended, it sets a flag to resume it early when it is next suspended.
+   */
+  void Resume()
```

Comment:
Should this function be thread-safe?

Reply:
Short answer: no.

Similar to the rest of the C++ language, low level building blocks shouldn't be threadsafe, unless their purpose is to actually be threads building blocks. And in the case of co-routines, as they are supposed to _replace_ threads, their purpose is to avoid threads altogether. The scheduler which is what will be responsible for resuming a coroutine will be either single-threaded, or thread-safe itself. It shouldn't be the responsability of the low level building block to make that decision, as it'd be a cost that's paid regardless of the intention: single-threaded or not.